IronPython3Extension/Extension.cs

```diff
@@ -29,9 +35,78 @@ public void Dispose()
         }
 
-        public void Ready(ReadyParams sp)
+        /// <summary>
+        /// Action to be invoked when the Dynamo has started up and is ready
+        /// for user interaction. 
+        /// </summary>
+        /// <param name="rp"></param>
+        public void Ready(ReadyParams rp)
+        {
+            var extraPath = Path.Combine(new FileInfo(Assembly.GetExecutingAssembly().Location).Directory.Parent.FullName, "extra");
+            var alc = new IsolatedPythonContext(Path.Combine(extraPath, $"{PythonEvaluatorAssembly}.dll"));
+            var dsIronAssem = alc.LoadFromAssemblyName(new AssemblyName(PythonEvaluatorAssembly));
+
+            //load the engine into Dynamo ourselves.
+            LoadPythonEngine(dsIronAssem);
+
+            //we used to do this:
+            //but it's not neccesary to load anything into the VM.
+            //instead we skip all the extra work and trigger the side effect we want
+            //which is re executing the graph after the dsIronPython evaluator is loaded into the PythonEngineManager.
+            //rp.StartupParams.LibraryLoader.LoadNodeLibrary(dsIronAssem);
+
+            if (rp.CurrentWorkspaceModel is HomeWorkspaceModel hwm)
+            {
+                foreach (var n in hwm.Nodes)
+                {
+                    n.MarkNodeAsModified(true);
+                }
+                hwm.Run();
+            }
+        }
+
+        private static void LoadPythonEngine(Assembly assembly)
         {
+            if (assembly == null)
+            {
+                return;
+            }
+
+            // Currently we are using try-catch to validate loaded assembly and Singleton Instance method exist
+            // but we can optimize by checking all loaded types against evaluators interface later
+            try
+            {
+                Type eType = null;
+                PropertyInfo instanceProp = null;
+                try
+                {
+                    eType = assembly.GetTypes().FirstOrDefault(x => typeof(PythonEngine).IsAssignableFrom(x) && !x.IsInterface && !x.IsAbstract);
+                    if (eType == null) return;
+
+                    instanceProp = eType?.GetProperty("Instance", BindingFlags.NonPublic | BindingFlags.Static);
+                    if (instanceProp == null) return;
+                }
+                catch
+                {
+                    // Ignore exceptions from iterating assembly types.
+                    return;
+                }
 
+                PythonEngine engine = (PythonEngine)instanceProp.GetValue(null);
+                if (engine == null)
+                {
+                    throw new Exception($"Could not get a valid PythonEngine instance by calling the {eType.Name}.Instance method");
+                }
+
+                if (PythonEngineManager.Instance.AvailableEngines.All(x => x.Name != engine.Name))
+                {
+                    PythonEngineManager.Instance.AvailableEngines.Add(engine);
```

Comment:
There are plans to make AvailableEngines internal and only expose a readonly version of it...
It seemed dangerous in D4R to have write access to the engine list...
Not sure anymore

Reply:
https://github.com/DynamoDS/Dynamo/compare/refactor_python_engines_api?expand=1

>

Reply:
I think it's fine if you go through the API break / obsolete dance as always, or release a new major version of Dynamo 😉 .

But in addition it sounded like you were suggesting adding public API for loading / injecting the engine? That does sound safer than having direct access to the collection.